sys/drv/rtl8139.c

```diff
@@ -32,18 +50,133 @@ static int rtl_reset(rtl8139_state_t *state) {
   return 1;
 }
 
-static int rtl8139_attach(device_t *dev) {
+static void do_receive(rtl8139_state_t *state) {
+  /* the device insert their own header which contains size and status of packet
+   */
+  uint16_t size = *(uint16_t *)(state->rx_buf + state->rx_offset + 2);
+
+  struct sk_buff *skb = kmalloc(M_DEV, sizeof(struct sk_buff), M_ZERO);
+  skb->length = size;
+  skb->data = kmalloc(M_DEV, size, M_ZERO);
+  memcpy(skb->data, (void *)(state->rx_buf + state->rx_offset + 4), size);
+  /* TODO: handle buffer overflow */
+  /* Enqueue */
+  /* TODO: WITH_MTX_LOCK { */
+  TAILQ_INSERT_TAIL(&state->rx_queue, skb, queue);
+
+  state->rx_offset += size + /* rtl8139 header */ 4;
+  bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
+}
+
+static void do_transceive(rtl8139_state_t *state) {
+  /* TODO */
+}
+
+static intr_filter_t rtl8139_intr(void *data) {
+  rtl8139_state_t *state = data;
+  uint16_t status = bus_read_2(state->regs, RL_ISR);
+
+  if (status & RL_ISR_RX_OK)
+    do_receive(state);
+  else if (status & RL_ISR_TX_OK)
+    do_transceive(state);
```

Comment:
Not sure, but I checked some implementations and it looks like they can be set at the same time, please check it. If not, what do you think about an assert?